pkg/distributor/distributor.go

```diff
@@ -732,6 +734,56 @@ func (d *Distributor) prePushForwardingMiddleware(next push.Func) push.Func {
 	}
 }
 
+// instanceLimitsMiddleware checks for instance limits and rejects request if this instance cannot process it at the moment.
+func (d *Distributor) instanceLimitsMiddleware(next push.Func) push.Func {
+	return func(ctx context.Context, req *mimirpb.WriteRequest, callerCleanup func()) (*mimirpb.WriteResponse, error) {
+		// We will report *this* request in the error too.
+		inflight := d.inflightPushRequests.Inc()
+		reqSize := int64(req.Size())
+		inflightBytes := d.inflightPushRequestsBytes.Add(reqSize)
+
+		// Decrement counter after all ingester calls have finished or been cancelled.
+		cleanup := func() {
+			callerCleanup()
+			d.inflightPushRequests.Dec()
+			d.inflightPushRequestsBytes.Sub(reqSize)
+		}
+		cleanupInDefer := true
+		defer func() {
+			if cleanupInDefer {
+				cleanup()
+			}
+		}()
+
+		userID, err := tenant.TenantID(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		span := opentracing.SpanFromContext(ctx)
+		if span != nil {
+			span.SetTag("organization", userID)
+		}
+
+		if d.cfg.InstanceLimits.MaxInflightPushRequests > 0 && inflight > int64(d.cfg.InstanceLimits.MaxInflightPushRequests) {
+			return nil, errMaxInflightRequestsReached
+		}
+
+		if d.cfg.InstanceLimits.MaxIngestionRate > 0 {
+			if rate := d.ingestionRate.Rate(); rate >= d.cfg.InstanceLimits.MaxIngestionRate {
+				return nil, errMaxIngestionRateReached
+			}
+		}
+
+		if d.cfg.InstanceLimits.MaxInflightPushRequestsBytes > 0 && inflightBytes > int64(d.cfg.InstanceLimits.MaxInflightPushRequestsBytes) {
+			return nil, errMaxInflightRequestsBytesReached
+		}
```

Comment:
At the beginning of this function we calculate the `inflight` number, then the `inflightBytes` number. In the deferred cleanup with also first update the `inflight` number, then the `inflightBytes` number. So just for the readability of this code it might be a bit easier if we switch around the rate limit and the `inflightBytes` limit instead of applying the rate limit in the middle, WDYT?:

```suggestion
		if d.cfg.InstanceLimits.MaxInflightPushRequests > 0 && inflight > int64(d.cfg.InstanceLimits.MaxInflightPushRequests) {
			return nil, errMaxInflightRequestsReached
		}

		if d.cfg.InstanceLimits.MaxInflightPushRequestsBytes > 0 && inflightBytes > int64(d.cfg.InstanceLimits.MaxInflightPushRequestsBytes) {
			return nil, errMaxInflightRequestsBytesReached
		}

		if d.cfg.InstanceLimits.MaxIngestionRate > 0 {
			if rate := d.ingestionRate.Rate(); rate >= d.cfg.InstanceLimits.MaxIngestionRate {
				return nil, errMaxIngestionRateReached
			}
		}
```

Alternatively, we could also move the rate limit to the very beginning of this function before we even increase `d.inflightPushRequests` and `d.inflightPushRequestsBytes`.

Reply:
> Alternatively, we could also move the rate limit to the very beginning of this function before we even increase d.inflightPushRequests and d.inflightPushRequestsBytes.

That's a good point, although it complicates cleanup a bit. But let me try.

Reply:
To keep the cleanup code simple, I just switched order to check rate as last, as you suggested.